kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/GenerateIcebergCompatActionUtils.java

```diff
@@ -0,0 +1,123 @@
+import io.delta.kernel.internal.TableConfig;
+import io.delta.kernel.internal.data.TransactionStateRow;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.icebergcompat.IcebergCompatV2MetadataValidatorAndUpdater;
+import io.delta.kernel.utils.CloseableIterable;
+import io.delta.kernel.utils.DataFileStatus;
+import java.net.URI;
+import java.util.Map;
+
+/** Utilities to convert Iceberg add/removes to Delta Kernel add/removes */
+public final class GenerateIcebergCompatActionUtils {
+
+  /**
+   * Create an add action {@link Row} that can be passed to {@link Transaction#commit(Engine,
+   * CloseableIterable)} from an Iceberg add.
+   *
+   * @param transactionState the transaction state from the built transaction
+   * @param fileStatus the file status to create the add with (contains path, time, size, and stats)
+   * @param partitionValues the partition values for the add
+   * @param dataChange whether or not the add constitutes a dataChange (i.e. append vs. compaction)
+   * @return add action row that can be committed to the transaction
+   * @throws UnsupportedOperationException if icebergWriterCompatV1 is not enabled
+   * @throws UnsupportedOperationException if maxRetries != 0 in the transaction
+   * @throws KernelException if stats are not present (required for icebergCompatV2)
+   * @throws UnsupportedOperationException if the table is partitioned (currently unsupported)
+   */
+  public static Row generateIcebergCompatWriterV1AddAction(
+      Row transactionState,
+      DataFileStatus fileStatus,
+      Map<String, Literal> partitionValues,
+      boolean dataChange) {
+    Map<String, String> config = getConfiguration(transactionState);
+
+    /* ----- Validate that this is a valid usage of this API ----- */
+    validateIcebergWriterCompatV1Enabled(config);
+    validateMaxRetriesSetToZero(transactionState);
+
+    /* ----- Validate this is valid write given the table's protocol & configurations ----- */
+    // note -- we know this must be enabled since IcebergCompatWriterV1 is enabled
+    if (TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(config)) {
+      // We require field `numRecords` when icebergCompatV2 is enabled
+      IcebergCompatV2MetadataValidatorAndUpdater.validateDataFileStatus(fileStatus);
+    }
+    // TODO possible stats enforcement for clustering (maybe not necessary due to icebergCompatV2?)
+
+    /* --- Validate and update partitionValues ---- */
+    // Currently we don't support partitioned tables; fail here
+    if (!getPartitionColumnsList(transactionState).isEmpty()) {
+      throw new UnsupportedOperationException(
+          "Currently GenerateIcebergCompatActionUtils "
+              + "is not supported for partitioned tables");
+    }
+    checkArgument(
+        partitionValues.isEmpty(), "Non-empty partitionValues provided for an unpartitioned table");
+
+    URI tableRoot = new Path(getTablePath(transactionState)).toUri();
```

Comment:
Can we please be consistent with our `TransactionStateRow.X` calls? i.e. `TransactionStateRow.getTablePath` ?